db/merge_test.cc

```diff
@@ -296,6 +298,87 @@ void testCounters(Counters& counters, DB* db, bool test_compaction) {
   }
 }
 
+void testCountersWithFlushAndCompaction(Counters& counters, DB* db) {
+  FlushOptions o;
+  o.wait = true;
+
+  // add some data, therefore we can trigger compaction
+  db->Put({}, "1", "1");
+  db->Flush(o);
+  db->Put({}, "2", "2");
+  db->Flush(o);
+  db->Put({}, "5", "5");
+  db->Flush(o);
+  db->Put({}, "6", "6");
+  db->Flush(o);
+
+  // wait flush finished
+  db->GetEnv()->SleepForMicroseconds(2000000);
+  //reinterpret_cast<DBImpl*>(db)->TEST_WaitForFlushMemTable(
+  //    db->DefaultColumnFamily());
+
+  std::shared_ptr<std::atomic<bool>> verified(new std::atomic<bool>(false));
+  std::atomic<int> cnt{0};
+  auto get_thread_id = [&cnt]() {
+    thread_local int thread_id{cnt++};
+    return thread_id;
+  };
+
+  rocksdb::SyncPoint::GetInstance()->SetCallBack(
+      "VersionSet::LogAndApply::WriterIsWaiting", [&, verified](void* arg) {
+        InstrumentedMutex* mtx = reinterpret_cast<InstrumentedMutex*>(arg);
+        auto thread_id = get_thread_id();
+        // condition variable is fine to be false-postive
+        // it is always legal to wake up and do nothing
+        while (!*verified) {
+          mtx->AssertHeld();
+          mtx->Unlock();
+          //std::cout << "thread - " + std::to_string(thread_id) +
+          //                 " is waiting\n";
+          db->GetEnv()->SleepForMicroseconds(1000000);
+          mtx->Lock();
+
+          // I am 100% sure the leader is thread - 0 and there are 2 working
+          // background threads
+          if (thread_id == 0 && cnt == 2) {
+            break;
+          }
+        }
+      });
+  rocksdb::SyncPoint::GetInstance()->EnableProcessing();
+
+  Slice begin("0");
+  Slice end("3");
+  db->SuggestCompactRange(db->DefaultColumnFamily(), &begin, &end);
+
+  // ensure the compaction job gets executed eariler than the incoming flush job
```

Comment:
Can you be more specific about the interleaving between flush and compaction using sync points?

Reply:
CC @JimChengLin

Reply:
@riversand963 done :)